presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayJoin.java

```diff
@@ -0,0 +1,191 @@
+import static com.facebook.presto.type.TypeUtils.readStructuralBlock;
+import static com.facebook.presto.util.Failures.checkCondition;
+import static com.facebook.presto.util.Reflection.methodHandle;
+import static java.lang.String.format;
+
+public final class ArrayJoin
+        extends ParametricScalar
+{
+    public static final ArrayJoin ARRAY_JOIN = new ArrayJoin();
+    public final ArrayJoinWithNullReplacement arrayJoinWithNullReplacement = new ArrayJoinWithNullReplacement();
+
+    private static final TypeSignature VARCHAR_TYPE_SIGNATURE = parseTypeSignature(StandardTypes.VARCHAR);
+    private static final String FUNCTION_NAME = "array_join";
+    private static final String DESCRIPTION = "Concatenates the elements of the given array using a delimiter and an optional string to replace nulls";
+    private static final Signature SIGNATURE = new Signature(FUNCTION_NAME, ImmutableList.of(typeParameter("T")), StandardTypes.VARCHAR,
+            ImmutableList.of("array<T>", StandardTypes.VARCHAR),
+            false, false);
+
+    private class ArrayJoinWithNullReplacement extends ParametricScalar
+    {
+        private final Signature signature = new Signature(FUNCTION_NAME, ImmutableList.of(typeParameter("T")), StandardTypes.VARCHAR,
+                ImmutableList.of("array<T>", StandardTypes.VARCHAR, StandardTypes.VARCHAR),
+                false, false);
+
+        @Override
+        public Signature getSignature()
+        {
+            return signature;
+        }
+
+        @Override
+        public boolean isHidden()
+        {
+            return false;
+        }
+
+        @Override
+        public boolean isDeterministic()
+        {
+            return true;
+        }
+
+        @Override
+        public String getDescription()
+        {
+            return DESCRIPTION;
+        }
+
+        @Override
+        public FunctionInfo specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+        {
+            return ArrayJoin.this.specializeArrayJoin(types, arity, typeManager, functionRegistry, ImmutableList.of(false, false, false));
+        }
+
+    }
+
+    @Override
+    public Signature getSignature()
+    {
+        return SIGNATURE;
+    }
+
+    @Override
+    public boolean isHidden()
+    {
+        return false;
+    }
+
+    @Override
+    public boolean isDeterministic()
+    {
+        return true;
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return DESCRIPTION;
+    }
+
+    @Override
+    public FunctionInfo specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        return specializeArrayJoin(types, arity, typeManager, functionRegistry, ImmutableList.of(false, false));
+    }
+
+    private FunctionInfo specializeArrayJoin(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry, List<Boolean> nullableArguments)
```

Comment:
Actually, I think you can just inline the value of those, since they return constants. I normally don't like doing that, because I think it's error-prone, but in this case I think it'll make the code nicer